src/pybind/rados/rados.pyx

```diff
@@ -2272,6 +2282,148 @@ class ReadOpCtx(ReadOp, OpCtx):
     """read operation context manager"""
 
 
+cdef void __watch_callback(void *_arg, int64_t _notify_id, uint64_t _cookie,
+                           uint64_t _notifier_id, void *_data,
+                           size_t _data_len) with gil:
+    """
+    Watch callback
+    """
+    cdef object watch = <object>_arg
+    data = None
+    if _data != NULL:
+        data = (<char *>_data)[:_data_len]
```

Comment:
Should this just use `decode_cstr`?

Reply:
I suppose it could be:
```
decode_cstr((<char *>_data)[:_data_len])
```
Because I don't think the message is a null terminated string in general case. And it would be nicer  when message is sent by the python notify: e.g. we would have in the python test (after sending 'test' message):
```
eq(self.notify_data[watch_id1], 'test')
```
instead of:
```
eq(self.notify_data[watch_id1], b'test')
```
But I am not sure what would `decode_cstr` do for messages generated by `c++` (usual case)? I don't know if the content of such messages will ever be useful for python as there is no a decoder... Anyway, I tried it and at least it did not crash when receiving "mirror enable/disable" notifications from librbd, so I can change it to `decode_cstr` if you think it is right.

Reply:
I added a commit that does `decode_cstr` for illustration. I will squash or drop it depending on if you like it or not.

Reply:
True -- I forgot that it's really just a blob of data. In that case, it should probably always return a `bytes` object.